[L0] Support for counter-based events using L0 driver

winstonzhang-intel opened the pull request:
Counter-based events implementation.
Counter-based events can be enabled via the flag UR_L0_USE_DRIVER_COUNTER_BASED_EVENTS=1

Below are the three conformance tests that ensures that this implementation is working as expected:
```
PASS: SYCL :: InorderQueue/in_order_buffs.cpp (1 of 1)

Testing Time: 6.09s

Total Discovered Tests: 1
  Passed: 1 (100.00%)

PASS: SYCL :: InorderQueue/in_order_kernels.cpp (1 of 1)

Testing Time: 6.29s

Total Discovered Tests: 1
  Passed: 1 (100.00%)


PASS: SYCL :: Basic/in_order_queue_status.cpp (1 of 1)

Testing Time: 7.55s

Total Discovered Tests: 1
  Passed: 1 (100.00%)
```
LLVM Draft with CI passing: https://github.com/intel/llvm/pull/12848
Rebased against Raiyan's in-order list [implementation](https://github.com/oneapi-src/unified-runtime/pull/1372/)

mabraham commented:
It's create to see things starting!

winstonzhang-intel commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1498,12 +1522,11 @@ ur_event_handle_t ur_queue_handle_t_::getEventFromQueueCache(bool IsMultiDevice,
+    ur_queue_handle_t Queue, ur_event_handle_t *Event, ur_command_t CommandType,
+    ur_command_list_ptr_t CommandList, bool IsInternal, bool IsMultiDevice,
+    std::optional<bool> HostVisible,
+    std::optional<bool> usingCounterBasedEvents) {
```
Good catch, let me fix that.

MichalMrozek reviewed (changes requested):
1. you can only use counter based events for in order queues, they do not work for out of order queues, you need to check sycl queue parameters instead of parsing over ze queues.
2. you cannot call reset on those events
3. why you disable fences ?
4. why you create all events with immediate and non immediate flags? Driver wouldn't be able to optimize this. Events should be selected basing on usage
5. why you add synchronous mode when counter based events are present ? 
6. event pools are per context which can have in order and out of order queues, you need to differentiate when you obtain events to not get counter-based event for ooq.
7. you cannot evaluate each time ur_queue_handle_t_::usingCounterBasedEvents() , this needs to be const flag set at queue creation, in order queue cannot be made out of order queue, map browsing would kill perf

winstonzhang-intel commented:
@MichalMrozek Thanks for the feedback.
1. you can only use counter based events for in order queues, they do not work for out of order queues, you need to check sycl queue parameters instead of parsing over ze queues.
I do make sure that we only use counter-based events for inOrderQueue. In additionally, I have to make sure that for the ZeQueues that are being used they have the ZE_COMMAND_QUEUE_FLAG_IN_ORDER flag
2. you cannot call reset on those events
Good catch, fixed
3. why you disable fences ?
My understanding was that with counter-based events, we no longer need barriers between the different lists, therefore, I removed the fences if we are using counter-based events.
4. why you create all events with immediate and non immediate flags? Driver wouldn't be able to optimize this. Events should be selected basing on usage
I agree, at first I was taking the L0 cts as example.  But come to think of it this makes sense.
5. why you add synchronous mode when counter based events are present?
Same as above.
6. event pools are per context which can have in order and out of order queues, you need to differentiate when you obtain events to not get counter-based event for ooq.
7. you cannot evaluate each time ur_queue_handle_t_::usingCounterBasedEvents() , this needs to be const flag set at queue creation, in order queue cannot be made out of order queue, map browsing would kill perf
Agree, fixed.

MichalMrozek commented on source/adapters/level_zero/event.cpp:
```diff
@@ -1071,14 +1072,18 @@ ur_result_t EventCreate(ur_context_handle_t Context, ur_queue_handle_t Queue,
   if (auto Res = Context->getFreeSlotInExistingOrNewPool(
-          ZeEventPool, Index, HostVisible, ProfilingEnabled, Device))
+          ZeEventPool, Index, HostVisible, ProfilingEnabled, Device,
+          CounterBasedEventEnabled.has_value()
```
can you use Boolean instead of std::optional ?

MichalMrozek commented on source/adapters/level_zero/event.cpp:
```diff
@@ -1124,8 +1129,8 @@ ur_result_t ur_event_handle_t_::reset() {
     HostVisibleEvent = nullptr;
-
-  ZE2UR_CALL(zeEventHostReset, (ZeEvent));
+  if (!UrQueue->usingCounterBasedEvents())
```
I think that event should know this instead of going to queue.

winstonzhang-intel replied:
Agreed, I can keep track of it during the creation of event. However, if it is passed in I would have no way of querying eventpool_desc.pnext from eventpool

MichalMrozek commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1364,6 +1374,27 @@ bool ur_queue_handle_t_::isInOrderQueue() const {
           0);
 }
 
+/*
```
please remove commented code

MichalMrozek commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1364,6 +1374,27 @@ bool ur_queue_handle_t_::isInOrderQueue() const {
+*/
+
+bool ur_queue_handle_t_::usingCounterBasedEvents() const {
+  /*
```
please remove commented code

MichalMrozek commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1818,6 +1850,9 @@ ur_queue_handle_t_::ur_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
   if (QueueIndex != 0) {
     ZeCommandQueueDesc.flags = ZE_COMMAND_QUEUE_FLAG_EXPLICIT_ONLY;
   }
+  if (Queue->usingCounterBasedEvents()) {
```
this should be other way around, if queue is in order , then you can create counter based events.

winstonzhang-intel replied:
ZeQueue hasn't been created at this point. If we are using counter-based event then the zequeue needs to reflect that.

raiyanla replied:
I believe @MichalMrozek is referring to the SYCL Queue being passed in AKA the UR Queue object in this line. You can check if this Queue was created to be in-order using the Queue->isInOrderQueue(). 

Also I believe it should also be noted that ZE_COMMAND_QUEUE_FLAG_IN_ORDER flag should only used to create immediate command lists specifically using zeCommandListCreateImmediate(), while I see this function is used to create an ordinary queue for regular command lists using zeCommandQueueCreate(). Please correct me if I'm wrong @MichalMrozek 

Reference at: https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=ze_command_queue_flag_in_order#_CPPv4N23ze_command_queue_flag_t30ZE_COMMAND_QUEUE_FLAG_IN_ORDERE

winstonzhang-intel replied:
@raiyanla Thanks for your input. Queue->usingCounterBasedEvents() does check for whether the Queue is in order with Queue->isInOrderQueue(). It also checks for ZE_COMMAND_QUEUE_FLAG_IN_ORDER, which I believe is also required for zeCommandQueueCreate(). See example in L0 coformance test [here](https://github.com/oneapi-src/level-zero-tests/blob/5990783066d8114e220ffd57061201e4f8d7cb17/conformance_tests/core/test_cmdlist/src/test_cmdlist.cpp#L823)

raiyanla replied:
@winstonzhang-intel Gotcha, yeah I just saw your usingCounterBasedEvents() already checks for isInOrderQueue() now. Thanks for linking that L0 conformance test, that is interesting, it seems to conflict with the specification definition of it above, since the spec link indicates its ZE_COMMAND_QUEUE_FLAG_IN_ORDER should be for immediate command lists only. @MichalMrozek Can you confirm which is is appropriate? Since the spec and conformance tests are conflicting.

MichalMrozek replied:
For regular  command lists the flag is ZE_COMMAND_LIST_FLAG_IN_ORDER , passed in command list descriptor
For immediate command lists the flag is ZE_COMMAND_QUEUE_FLAG_IN_ORDER , passed in queue descriptor

MichalMrozek commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1854,32 +1889,38 @@ ur_result_t ur_queue_handle_t_::createCommandList(
 
-  ZE2UR_CALL(zeFenceCreate, (ZeCommandQueue, &ZeFenceDesc, &ZeFence));
+  if (!usingCounterBasedEvents()) {
+    ZE2UR_CALL(zeFenceCreate, (ZeCommandQueue, &ZeFenceDesc, &ZeFence));
```
do not disable fences, they are still needed to recycle command lists.

MichalMrozek reviewed (changes requested):
It looks like this change has dependency on in order support

pbalcer commented on source/adapters/level_zero/event.hpp:
```diff
@@ -29,9 +29,10 @@
+ur_result_t
+EventCreate(ur_context_handle_t Context, ur_queue_handle_t Queue,
+            bool IsMultiDevice, bool HostVisible, ur_event_handle_t *RetEvent,
+            std::optional<bool> CounterBasedEventEnabled = std::nullopt);
```
```suggestion
            bool CounterBasedEventEnabled = false);
```

?

pbalcer commented on source/adapters/level_zero/context.cpp:
```diff
@@ -468,7 +468,8 @@ static const uint32_t MaxNumEventsPerPool = [] {
 
 ur_result_t ur_context_handle_t_::getFreeSlotInExistingOrNewPool(
```
there's ought to be a better way of allocating an event that doesn't require passing around half a dozen arguments to multiple functions. Maybe some sort of flag?

`auto event = Pool->allocateEvent(HOST_VISIBLE | ENABLE_PROFILER | COUNTER_BASED);`

having functions that accept multiple boolean values is error prone.

winstonzhang-intel replied:
@pbalcer Thanks for the feedback! We could create a new enum just for this case. However, this seems to be a bit overkill just for this one function. It is only used once event.cpp during event creation.

pbalcer replied:
It's not just this function. All the various boolean event parameters are being passed around in multiple functions. Just from a quick search:
`createEventAndAssociateQueue, getEventFromQueueCache, EventCreate, getEventFromContextCache, getFreeSlotInExistingOrNewPool, getZeEventPoolCache`

winstonzhang-intel replied:
@pbalcer This is quite a large refactoring. Functions with these booleans are abundant. Additionally, it would make sense to add this in **ur_api**, however, we would need to add this in spec before I can do so. Perhaps I can file a new ticket and have this be its own patch?

pbalcer replied:
Sure, this was just a suggestion.

pbalcer commented on source/adapters/level_zero/context.cpp:
```diff
@@ -510,6 +511,12 @@ ur_result_t ur_context_handle_t_::getFreeSlotInExistingOrNewPool(
       ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_KERNEL_TIMESTAMP;
+    if (CounterBasedEventEnabled) {
+      ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
+      ze_event_pool_counter_based_exp_desc_t counterBasedExt = {
```
I think you might need to update the `getZeEventPoolCache(HostVisible, ProfilingEnabled, ZeDevice);` call above. Otherwise, based on what the queue was that happened to call this function first for this combination of parameters, the event pool might or might not have this flag set.

And, again, I suggest a small refactoring here to use flags (or some other mechanism) instead of passing yet another boolean.

winstonzhang-intel replied:
The Queue should have the flag set on creation. Because of that, all the subsequent eventpools should be cached with counter-based events enabled.

pbalcer replied:
But this cache is in the context, so if you have two queues, one with counter-based events enabled and one without, the one that was first in this function will create the event pool.

MichalMrozek replied:
event pools are stored in context not in a queue
There may be different types of queues coming in and they would require different type of events.
Pool cannot return incompatible event.

winstonzhang-intel replied:
Using the same mechanism as previously implemented, the eventpool cache returned will be guaranteed to be with counterbased enabled if it is selected, since now there is a portion of the eventpoolcache that is reserved just for counterbased event pools.

pbalcer replied:
👍 this desperately needs a refactor though :)

winstonzhang-intel replied:
I'll file a ticket tomorrow for all the refactoring, should be pretty straight forward 👍

winstonzhang-intel commented on source/adapters/level_zero/context.cpp:
```diff
@@ -468,7 +468,8 @@ static const uint32_t MaxNumEventsPerPool = [] {
 
 ur_result_t ur_context_handle_t_::getFreeSlotInExistingOrNewPool(
```
Fixed. There will only be regular cmdlist here, immediate cmdlist is retrieved via &ur_queue_handle_t_::ur_queue_group_t::getImmCmdList() in queue.cpp

winstonzhang-intel replied:
good catch, fixed

winstonzhang-intel commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1983,8 +2008,10 @@ ur_command_list_ptr_t &ur_queue_handle_t_::ur_queue_group_t::getImmCmdList() {
-  if (QueueIndex != 0) {
+  if (Queue->Device->useDriverInOrderLists() && Queue->isInOrderQueue()) {
+    ZeCommandQueueDesc.flags = ZE_COMMAND_QUEUE_FLAG_IN_ORDER;
+  } else if (QueueIndex != 0) {
```
This is part of the in-order list patch `https://github.com/oneapi-src/unified-runtime/pull/1372/.` Not related to counter-based events.

MichalMrozek replied:
can you re-base ?
 the in order patch is merged and has different change here

winstonzhang-intel commented on source/adapters/level_zero/context.hpp:
```diff
@@ -141,9 +141,9 @@ struct ur_context_handle_t_ : _ur_object {
+  std::vector<std::list<ze_event_pool_handle_t>> ZeEventPoolCache{8};
   std::vector<std::unordered_map<ze_device_handle_t, size_t>>
-      ZeEventPoolCacheDeviceMap{4};
+      ZeEventPoolCacheDeviceMap{8};
```
2 EventPoolCacheDeviceMap reserved for each: hostVisible, not hostvisible, device and hostvisible and device and not hostvisible. Since we can't query the ext flags of eventpooldesc from eventpoolhandle, we have to reserve the space in case of each of these situations.

MichalMrozek replied:
What if you need profiling on top of that?
Which pool would you choose?

If you want to permute all combination this would quickly be out of control.
- host visible
- counter based for immediate
- counter based for non immediate
- profiling

we are already at 16

winstonzhang-intel commented on source/adapters/level_zero/context.hpp:
```diff
@@ -206,35 +207,33 @@ struct ur_context_handle_t_ : _ur_object {
-      } else {
-        return WithProfiling ? &ZeEventPoolCache[0] : &ZeEventPoolCache[1];
-      }
+      profiling_index_a = 0;
```
yes to choose the cache index depending on hostVisible and CounterBasedEventEnabled.

MichalMrozek commented on source/adapters/level_zero/context.hpp:
```diff
@@ -206,36 +208,43 @@ struct ur_context_handle_t_ : _ur_object {
+                                  int *profiling_index_a,
+                                  int *profiling_index_b) {
+    if (HostVisible) {
+      *profiling_index_a = 0;
```
can you make this code more maintainable ?
assign some enum for each pool type, shift by enum sizes ?

Right now it would be very difficult to understand what is happening here and why, also profiling_index_a and profiling_index_b doesn't tell a lot what happens here.

MichalMrozek reviewed (commented):
please re-base patch after in order command list change, right now it is hard to distinguish what are new additions and what comes from in order lists patch.

nrspruit commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -1861,19 +1877,18 @@ ur_result_t ur_queue_handle_t_::createCommandList(
 
   uint32_t QueueGroupOrdinal;
+  ZeStruct<ze_command_list_desc_t> ZeCommandListDesc;
+  if (usingCounterBasedEvents()) {
```
Does "using counter based events" override the check for:
if (Device->useDriverInOrderLists() && isInOrderQueue()) ?

Or should we be checking for all of them?

nrspruit replied:
nvm, I found the code ie:
  this->counterBasedEventsEnabled = isInOrderQueue() &&
                                    Device->useDriverInOrderLists() &&
                                    useDriverCounterBasedEvents;
                                    
  please disregard.

winstonzhang-intel replied:
Missed this during my rebase. But when counterbased is enabled then both inorder list and inorder queue will be enabled. Because of this we can just check for Device->useDriverInOrderLists() && isInOrderQueue()

nrspruit reviewed (approved):
LGTM, @MichalMrozek , please check to see if you have any further comments or issues with these changes.

MichalMrozek commented on source/adapters/level_zero/context.cpp:
```diff
@@ -510,6 +511,18 @@ ur_result_t ur_context_handle_t_::getFreeSlotInExistingOrNewPool(
     if (ProfilingEnabled)
       ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_KERNEL_TIMESTAMP;
+    if (CounterBasedEventEnabled) {
+      ZeEventPoolDesc.flags |= ZE_EVENT_POOL_FLAG_HOST_VISIBLE;
```
why HOST_VISIBLE even if HostVisible is not set?

winstonzhang-intel replied:
removed

MichalMrozek commented on source/adapters/level_zero/context.hpp:
```diff
@@ -204,38 +206,49 @@ struct ur_context_handle_t_ : _ur_object {
+    }
+  }
+
+  ur_result_t calculateCacheIndex(bool HostVisible,
```
how do you differentiate between counter based events for immediate and regular command lists?

winstonzhang-intel replied:
added immediate detection

MichalMrozek commented on source/adapters/level_zero/context.hpp:
```diff
@@ -204,38 +206,49 @@ struct ur_context_handle_t_ : _ur_object {
-        return WithProfiling ? &ZeEventPoolCache[2] : &ZeEventPoolCache[3];
-      }
+      return WithProfiling ? &ZeEventPoolCache[CacheType]
+                           : &ZeEventPoolCache[CacheType + 1];
```
shouldn't you multiply by 2? 

Cache type +1 gives next enum value
for HostVisibleCacheType you will get HostInvisibleCacheType

winstonzhang-intel replied:
+1 is for with profiling and without profiling for each hostvisible/!hostvisible pair for each regular, regular with counterbase and imm with counterbase.

MichalMrozek replied:
But you will get cache types as values from 0..5
So +1 values would be in range 1..6
1. Code would never use 12 cache entries.
2. Code would incorrectly overlap some entries

winstonzhang-intel replied:
Oversight on my part, fixed now. Thanks.

nrspruit commented:
> 1. you can only use counter based events for in order queues, they do not work for out of order queues, you need to check sycl queue parameters instead of parsing over ze queues.
> 2. you cannot call reset on those events
> 3. why you disable fences ?
> 4. why you create all events with immediate and non immediate flags? Driver wouldn't be able to optimize this. Events should be selected basing on usage
> 5. why you add synchronous mode when counter based events are present ?
> 6. event pools are per context which can have in order and out of order queues, you need to differentiate when you obtain events to not get counter-based event for ooq.
> 7. you cannot evaluate each time ur_queue_handle_t_::usingCounterBasedEvents() , this needs to be const flag set at queue creation, in order queue cannot be made out of order queue, map browsing would kill perf

So, the CI failure here:
urUSMHostAllocTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics

is because the reused counting event is coming up as signaled so it immediately completes. Either this is a bug in the L0 gpu driver if the spec says we cannot reset or we are missing something in this implementation.

nrspruit commented:
> > 1. you can only use counter based events for in order queues, they do not work for out of order queues, you need to check sycl queue parameters instead of parsing over ze queues.
> > 2. you cannot call reset on those events
> > 3. why you disable fences ?
> > 4. why you create all events with immediate and non immediate flags? Driver wouldn't be able to optimize this. Events should be selected basing on usage
> > 5. why you add synchronous mode when counter based events are present ?
> > 6. event pools are per context which can have in order and out of order queues, you need to differentiate when you obtain events to not get counter-based event for ooq.
> > 7. you cannot evaluate each time ur_queue_handle_t_::usingCounterBasedEvents() , this needs to be const flag set at queue creation, in order queue cannot be made out of order queue, map browsing would kill perf
> 
> So, the CI failure here: urUSMHostAllocTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics
> 
> is because the reused counting event is coming up as signaled so it immediately completes. Either this is a bug in the L0 gpu driver if the spec says we cannot reset or we are missing something in this implementation.

Issue is fixed in the latest patch, the issue was the desc was not being used.

winstonzhang-intel commented:
> > > 1. you can only use counter based events for in order queues, they do not work for out of order queues, you need to check sycl queue parameters instead of parsing over ze queues.
> > > 2. you cannot call reset on those events
> > > 3. why you disable fences ?
> > > 4. why you create all events with immediate and non immediate flags? Driver wouldn't be able to optimize this. Events should be selected basing on usage
> > > 5. why you add synchronous mode when counter based events are present ?
> > > 6. event pools are per context which can have in order and out of order queues, you need to differentiate when you obtain events to not get counter-based event for ooq.
> > > 7. you cannot evaluate each time ur_queue_handle_t_::usingCounterBasedEvents() , this needs to be const flag set at queue creation, in order queue cannot be made out of order queue, map browsing would kill perf
> > 
> > 
> > So, the CI failure here: urUSMHostAllocTest.Success/Intel_R__oneAPI_Unified_Runtime_over_Level_Zero___Intel_R__Arc_TM__A750_Graphics
> > is because the reused counting event is coming up as signaled so it immediately completes. Either this is a bug in the L0 gpu driver if the spec says we cannot reset or we are missing something in this implementation.
> 
> Issue is fixed in the latest patch, the issue was the desc was not being used.

Thank you @nrspruit for looking into this. Once this patch is merged I will post another for the refactoring that @pbalcer requested.

EwanC commented on source/adapters/level_zero/queue.hpp:
```diff
@@ -234,6 +234,9 @@ struct ur_queue_handle_t_ : _ur_object {
   ur_queue_flags_t Properties;
 
+  // Keeps track of whether we are using Counter-based Events
+  bool counterBasedEventsEnabled = false;
```
nit: variable naming not quite consistent
```suggestion
  bool CounterBasedEventsEnabled = false;
```

EwanC commented on source/adapters/level_zero/queue.cpp:
```diff
@@ -955,6 +955,17 @@ ur_queue_handle_t_::ur_queue_handle_t_(
       ZeCommandListBatchComputeConfig.startSize();
   CopyCommandBatch.QueueBatchSize = ZeCommandListBatchCopyConfig.startSize();
+
+  static const bool useDriverCounterBasedEvents = [] {
```
defining a lambda that only has one call site feels like extra complexity for little benefit.

nrspruit replied:
When it comes to querying environment variables, we define a lambda in all uses in the code. This provides consistency even if right now this is only used once.

EwanC commented on source/adapters/level_zero/event.hpp:
```diff
@@ -222,6 +223,11 @@ struct ur_event_handle_t_ : _ur_object {
+  // Keeps track of whether we are using Counter-based Events.
+  bool CounterBasedEventsEnabled = false;
+
+  bool usingCounterBasedEvents() const { return CounterBasedEventsEnabled; }
```
`CounterBasedEventsEnabled` looks like a public struct member, what's the benefit of defining a getter function to return it?

EwanC commented on source/adapters/level_zero/command_buffer.cpp:
```diff
@@ -933,7 +933,8 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
   }
-  if (MustSignalWaitEvent) {
+  if (MustSignalWaitEvent &&
+      !CommandBuffer->WaitEvent->CounterBasedEventsEnabled) {
```
If we didn't set the `CounterBasedEventEnabled` parameter to true when `WaitEvent` is [created](https://github.com/oneapi-src/unified-runtime/blob/4ef258c85ea39c5c5b4269ee9ff1768a0d11192b/source/adapters/level_zero/command_buffer.cpp#L438)
```cpp
UR_CALL(EventCreate(Context, nullptr, false, false,
                      &RetCommandBuffer->WaitEvent));
```
Why do we need to check here that `CounterBasedEventsEnabled` isn't set? Can this property of the event change after the event has been created?

nrspruit reviewed (approved):
LGTM with the updated changes.
